File: Request.java
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.HttpClient.Redirect;
import java.net.http.HttpResponse.BodyHandlers;
import java.time.Duration;
import java.util.ArrayList;
import java.util.concurrent.CompletableFuture;

public class Request {
    static final int NUM_REQUESTS = 100;
    static boolean bad = false;
    static boolean asyncBad = false;

    public static void main(String[] args) throws IOException, InterruptedException {
        HttpClient client = HttpClient.newBuilder().followRedirects(Redirect.NORMAL).build();

        long startTime = System.currentTimeMillis();
        ArrayList<CompletableFuture<Void>> futures = new ArrayList<>(NUM_REQUESTS);
        for (int i = 0; i < NUM_REQUESTS; i++) {
            HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create("https://www.example.com"))
                // .setHeader("header name", "value")
                // .POST(HttpRequest.BodyPublishers.ofString("request body"))
                .timeout(Duration.ofSeconds(1))
                .build();
            CompletableFuture<Void> future = client.sendAsync(request, BodyHandlers.ofString())
                // .thenApply(response -> {
                //     String body = response.body();
                //     System.out.println(body);
                //     String contentType = response.headers().firstValue("Content-Type").orElse(null);
                //     System.out.println(contentType);
                //     return response;
                // })
                .thenApply(HttpResponse::statusCode).thenAccept(statusCode -> {
                    if (statusCode != 200) {
                        asyncBad = true;
                    }
                });
            futures.add(future);
        }
        CompletableFuture.allOf(futures.toArray(new CompletableFuture<?>[0])).join();
        long endTime = System.currentTimeMillis();
        double elapsedTime = (endTime - startTime) / 1000.0;
        if (asyncBad) {
            System.out.println("Had failed asynchronous request");
        } else {
            System.out.println("Asynchronous elapsed time for " + NUM_REQUESTS + " requests = " + elapsedTime + " seconds");
        }

        startTime = System.currentTimeMillis();
        for (int i = 0; i < NUM_REQUESTS; i++) {
            HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create("https://www.example.com"))
                // .setHeader("header-name", "value")
                // .POST(HttpRequest.BodyPublishers.ofString("request body"))
                .timeout(Duration.ofSeconds(1))
                .build();
            HttpResponse<String> response = client.send(request, BodyHandlers.ofString());
            // String body = response.body();
            // System.out.println(body);
            // String contentType = response.headers().firstValue("Content-Type").orElse(null);
            // System.out.println(contentType);
            if (response.statusCode() != 200) {
                bad = true;
            }
        }
        endTime = System.currentTimeMillis();
        elapsedTime = (endTime - startTime) / 1000.0;
        if (bad) {
            System.out.println("Had failed synchronous request");
        } else {
            System.out.println(" Synchronous elapsed time for " + NUM_REQUESTS + " requests = " + elapsedTime + " seconds");
        }
    }
}

File: request.rs
use std::time::{self, Duration};

use reqwest::{
    Client,
    blocking,
    // header::{CONTENT_TYPE, HeaderMap, HeaderValue},
};

const NUM_REQUESTS: i32 = 10;

fn sync_main() {
    let mut bad = false;
    let client = blocking::Client::builder()
        .timeout(Duration::from_secs(1))
        .build()
        .unwrap();
    // let mut headers = HeaderMap::new();
    // headers.insert("header-name", HeaderValue::from_static("value"));
    // let client = blocking::Client::builder()
    //     .timeout(Duration::from_secs(1))
    //     .default_headers(headers)
    //     .build()
    //     .unwrap();
    let start_time = time::SystemTime::now();
    for _ in 0..NUM_REQUESTS {
        // let response = client
        //     .post("https://www.example.com")
        //     .body("request body")
        //     .send()
        //     .unwrap();
        let response = client.get("https://www.example.com").send().unwrap();
        if !response.status().is_success() {
            // code is not in range 200-299
            bad = true;
        } else {
            // let value = response.headers().get(CONTENT_TYPE);
            // let content_type = match value {
            //     Some(val) => val.to_str().unwrap(),
            //     None => "",
            // };
            // println!("{content_type}");
            // let body = response.text().unwrap();
            // println!("{body}")
        }
    }
    let end_time = time::SystemTime::now();
    let elapsed_time = end_time.duration_since(start_time).unwrap().as_secs_f64();
    if bad {
        println!("Had failed synchronous request");
    } else {
        println!(" Synchronous elapsed time for {NUM_REQUESTS} requests = {elapsed_time} seconds");
    }
}

#[tokio::main]
async fn main() {
    let mut futures = vec![];
    let mut async_bad = false;
    let client = Client::builder()
        .timeout(Duration::from_secs(1))
        .build()
        .unwrap();
    // let mut headers = HeaderMap::new();
    // headers.insert("header-name", HeaderValue::from_static("value"));
    // let client = Client::builder()
    //     .timeout(Duration::from_secs(1))
    //     .default_headers(headers)
    //     .build()
    //     .unwrap();
    let start_time = time::SystemTime::now();
    for _ in 0..NUM_REQUESTS {
        let local_client = client.clone();
        let future = tokio::spawn(async move {
            // let response = local_client
            //     .post("https://www.example.com")
            //     .body("request body")
            //     .send()
            //     .await
            //     .unwrap();
            let response = local_client
                .get("https://www.example.com")
                .send()
                .await
                .unwrap();
            if !response.status().is_success() {
                // code is not in range 200-299
                true
            } else {
                // let value = response.headers().get(CONTENT_TYPE);
                // let content_type = match value {
                //     Some(val) => val.to_str().unwrap(),
                //     None => "",
                // };
                // println!("{content_type}");
                // let body = response.text().await.unwrap();
                // println!("{body}");
                false
            }
        });
        futures.push(future);
    }
    for future in futures {
        if future.await.unwrap() {
            async_bad = true;
            break;
        }
    }
    let end_time = time::SystemTime::now();
    let elapsed_time = end_time.duration_since(start_time).unwrap().as_secs_f64();
    if async_bad {
        println!("Had failed asynchronous request");
    } else {
        println!("Asynchronous elapsed time for {NUM_REQUESTS} requests = {elapsed_time} seconds");
    }

    tokio::task::spawn_blocking(|| {
        sync_main();
    })
    .await
    .unwrap();
}
